lib/location/method_cloud_location.c

```diff
@@ -0,0 +1,244 @@
+#include "scan_cellular.h"
+#include "scan_wifi.h"
+#include "cloud_service/cloud_service.h"
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+BUILD_ASSERT(
+	IS_ENABLED(CONFIG_LOCATION_SERVICE_NRF_CLOUD) ||
+	IS_ENABLED(CONFIG_LOCATION_SERVICE_HERE) ||
+	IS_ENABLED(CONFIG_LOCATION_SERVICE_EXTERNAL),
+	"At least one location service, or handling the service externally must be enabled");
+
+/* Common for both */
+struct method_cloud_location_start_work_args {
+	struct k_work work_item;
+	const struct location_wifi_config *wifi_config;
+	const struct location_cellular_config *cell_config;
+	int64_t starting_uptime_ms;
+};
+
+static struct method_cloud_location_start_work_args method_cloud_location_start_work;
+static bool running;
+
+static void method_cloud_location_positioning_work_fn(struct k_work *work)
+{
+	struct method_cloud_location_start_work_args *work_data =
+		CONTAINER_OF(work, struct method_cloud_location_start_work_args, work_item);
+	const struct location_wifi_config *wifi_config = work_data->wifi_config;
+	const struct location_cellular_config *cell_config = work_data->cell_config;
+	struct wifi_scan_info *scan_wifi_info = NULL;
+	struct lte_lc_cells_info *scan_cellular_info = NULL;
+	int64_t scan_start_time;
+	int32_t used_timeout_ms;
+	int err = 0;
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	struct k_sem wifi_scan_ready;
+
+	k_sem_init(&wifi_scan_ready, 0, 1);
+#endif
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	struct k_sem ncellmeas_ready;
+
+	k_sem_init(&ncellmeas_ready, 0, 1);
+#endif
+
+	if (wifi_config != NULL && cell_config != NULL) {
+		used_timeout_ms = MIN(cell_config->timeout, wifi_config->timeout);
+	} else if (cell_config != NULL) {
+		used_timeout_ms = cell_config->timeout;
+	} else {
+		__ASSERT_NO_MSG(wifi_config != NULL);
+		used_timeout_ms = wifi_config->timeout;
+	}
+
+	location_core_timer_start(used_timeout_ms);
+	scan_start_time = k_uptime_get();
+
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	if (wifi_config) {
+		err = scan_wifi_start(&wifi_scan_ready);
+	} else {
+		k_sem_give(&wifi_scan_ready);
+	}
+#endif
+
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	if (cell_config) {
+		err = scan_cellular_start(&ncellmeas_ready);
+	} else {
+		k_sem_give(&ncellmeas_ready);
+	}
+#endif
+
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	k_sem_take(&wifi_scan_ready, K_FOREVER);
+	scan_wifi_info = scan_wifi_results_get();
+#endif
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	k_sem_take(&ncellmeas_ready, K_FOREVER);
+	scan_cellular_info = scan_cellular_results_get();
+#endif
+	if (!running) {
+		goto end;
+	}
+
+	location_core_timer_stop();
+
+	if (scan_cellular_info == NULL && scan_wifi_info == NULL) {
+		LOG_WRN("No cellular neighbor cells or Wi-Fi access points found");
+		err = -ENODATA;
+		goto end;
+	}
+
+#if defined(CONFIG_LOCATION_SERVICE_EXTERNAL)
+	struct location_data_cloud request = {
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+		.cell_data = (cell_config != NULL) ? scan_cellular_info : NULL,
```

Comment:
Not a big deal, but is it necessary to check if `cell_config` is not `NULL`? If it is, `scan_cellular_info` should also be `NULL`, I think. The same check is done twice for `cell_config` and twice for `wifi_config`.

Reply:
I've had so many versions of this file trying to clean flagging and how cell_config and scan_cellular_info are handled. Seems the end result wasn't clean :)

I've changed this and similar code few lines below for non-external service.